language/resource-api/README.md

```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
 ### Provider feature: `remote_resource`
 
 ```ruby
+# lib/puppet/type/nx9k_vlan.rb
 Puppet::ResourceApi.register_type(
   name: 'nx9k_vlan',
   features: [ 'remote_resource' ],
 )
 
-require 'puppet/util/network_device/simple/device'
-module Puppet::Util::NetworkDevice::Nexus
-  class Device < Puppet::Util::NetworkDevice::Simple::Device
-    def facts
-      # access the device and return facts hash
-    end
-  end
-end
-
-class Puppet::Provider::Nx9k_vlan::Nx9k_vlan
+# lib/puppet/provider/nx9k_vlan/nexus.rb
+class Puppet::Provider::Nx9k_vlan::Nexus
   def set(context, changes, noop: false)
     changes.each do |name, change|
       is = change.has_key? :is ? change[:is] : get_single(name)
       should = change[:should]
       # ...
-      context.device.do_something unless noop
+      if context.transport.verify
+        context.transport.do_something unless noop
+      else
+        raise Puppet::Transport::Error('Connection to remote target is not available')
+      end
     end
   end
 end
 ```
 
-Declaring this feature restricts the resource from being run "locally". It is expected to execute all external interactions through the `context.device` instance. The way that instance is set up is runtime specific. In Puppet, it is configured through the [`device.conf`](https://puppet.com/docs/puppet/5.3/config_file_device.html) file, and only available when running under [`puppet device`](https://puppet.com/docs/puppet/5.3/man/device.html). It is recommended to use `Puppet::Util::NetworkDevice::Simple::Device` as the base class for all devices, which automatically loads a configuration from the local filesystem of the proxy node where it is running on.
+Declaring this feature restricts the resource from being run "locally". It is expected to execute all external interactions through the `context.transport` instance. The way that instance is set up is runtime specific. In Puppet, it is configured through the [`device.conf`](https://puppet.com/docs/puppet/5.3/config_file_device.html) file, and only available when running under [`puppet device`](https://puppet.com/docs/puppet/5.3/man/device.html).
+
+### Transport
+
+```ruby
+# lib/puppet/transport/nexus_schema.rb
+Puppet::ResourceAPI.register_transport(
+  name: 'nexus', # points at class Puppet::Transport::Nexus
+  desc: 'Connects to a Cisco Nexus device'
+  features: [], # TODO
+  connection_info: {
+    connector_type: {
+      type: 'Enum[nxapi, ssh, grpc, pidgeon]',
+      desc: 'Which method to use to connect to the host'
+    },
+    hostname: {
+      type: 'String',
+      desc: 'The host to connect to.',
+    },
+    username: {
+      type: 'String',
+      desc: 'The user.',
+    },
+    password: {
+      type: 'Sensitive[String]',
```

Comment:
We discussed using sensitive like this in the task spec and decided it wasn't what we wanted. This means the password field will only accept a value of type Senstitive where what we mean is this value should be treated as sensitive information. For the task spec we instead added a new key `sensitive` next to type and description to capture this. I would encourage doing something similar here.

Reply:
the run_task function turns them into instances of Puppet::Sensitive and then the transports unwrap the sensitive values at the last minute.

https://github.com/puppetlabs/bolt/blob/fb186f9165103b399a6960a7fc42983c4ce714dd/bolt-modules/boltlib/lib/puppet/functions/run_task.rb#L135
https://github.com/puppetlabs/bolt/blob/fb186f9165103b399a6960a7fc42983c4ce714dd/lib/bolt/transport/ssh.rb#L110

Reply:
Added language explaining this.